Dart/src/com/jetbrains/lang/dart/ide/formatter/DartPreFormatProcessor.java

```diff
@@ -0,0 +1,133 @@
+package com.jetbrains.lang.dart.ide.formatter;
+
+import com.intellij.lang.ASTNode;
+import com.intellij.openapi.editor.Document;
+import com.intellij.openapi.editor.RangeMarker;
+import com.intellij.openapi.module.Module;
+import com.intellij.openapi.project.Project;
+import com.intellij.openapi.roots.ContentEntry;
+import com.intellij.openapi.roots.ModuleRootManager;
+import com.intellij.openapi.util.Key;
+import com.intellij.openapi.util.TextRange;
+import com.intellij.openapi.util.io.FileUtil;
+import com.intellij.openapi.vfs.VfsUtilCore;
+import com.intellij.openapi.vfs.VirtualFile;
+import com.intellij.psi.FileViewProvider;
+import com.intellij.psi.PsiDocumentManager;
+import com.intellij.psi.PsiElement;
+import com.intellij.psi.PsiFile;
+import com.intellij.psi.codeStyle.CodeStyleSettings;
+import com.intellij.psi.codeStyle.CodeStyleSettingsManager;
+import com.intellij.psi.impl.source.codeStyle.PreFormatProcessor;
+import com.jetbrains.lang.dart.DartLanguage;
+import com.jetbrains.lang.dart.analyzer.DartAnalysisServerService;
+import com.jetbrains.lang.dart.sdk.DartSdk;
+import com.jetbrains.lang.dart.sdk.DartSdkGlobalLibUtil;
+import org.dartlang.analysis.server.protocol.SourceEdit;
+import org.jetbrains.annotations.NotNull;
+
+import java.io.*;
+import java.util.Collection;
+import java.util.List;
+
+public class DartPreFormatProcessor implements PreFormatProcessor {
+  public static Key<Boolean> FORMAT_MARK = new Key<Boolean>("FORMAT_MARK");
+  public static Boolean FORMAT_MARKER = Boolean.TRUE;
+
+  private CodeStyleSettings mySettings;
+  private PsiFile myFile;
+  private Project myProject;
+  private int selectionOffset;
+
+  @NotNull
+  @Override
+  public TextRange process(@NotNull ASTNode element, @NotNull TextRange range) {
+    // TODO return range if not in a paste operation
+    final PsiElement psiElement = element.getPsi();
+    if (psiElement == null) return range;
+
+    if (!psiElement.getLanguage().is(DartLanguage.INSTANCE)) return range;
+
+    if (!psiElement.isValid()) return range;
+    myFile = psiElement.getContainingFile();
+    if (myFile == null || !myFile.isWritable()) return range;
+
+    myProject = psiElement.getProject();
+    mySettings = CodeStyleSettingsManager.getSettings(myProject);
+    final FileViewProvider viewProvider = myFile.getViewProvider();
+    final Document document = viewProvider.getDocument();
+    if (document == null) return range;
+
+    final String code = reformat(document.getText(), range);
+    if (code == null) return range;
+    final RangeMarker marker = document.createRangeMarker(range.getStartOffset(), range.getEndOffset(), true);
+    document.replaceString(0, document.getTextLength(), code);
```

Comment:
I'm afraid this RangeMarker can't survive with any reasonable range after total document text replacement.
The task to keep range after reformat doesn't look trivial. Probably PsiElementPointers can help, after reformatting all PsiElementPointers should survive. Althouh PsiElements change on reparse, in general PSI tree remains the same (because only whitespaces are changed), so this is where PsiElementPointer can survive and provide updated range. That's just a guess.